### ISSUE TITLE
Multiple JSF Applications found on same ClassLoader.  Unable to safely determine which FactoryManager instance to use. Defaulting to first match - When running more Virtual servers PAYARA-2613

### ISSUE THREAD

**Issue by rds76** [payara/Payara#1133]
Log file continually flooding with warning:

```
[2016-10-12T15:43:57.294+0200] [Payara 4.1] [WARNING] [] [javax.faces] [tid: _ThreadID=48 _ThreadName=http-listener-2(7)] [[
  Multiple JSF Applications found on same ClassLoader.  Unable to safely determine which FactoryManager instance to use. Defaulting to first match.]]
```

**Comment by mikecroft** [comment]
Can you give a bit more detail on your environment and how its configured? Have you turned off class loader delegation?

**Comment by OndrejM** [comment]
What do you mean "When running more virtual servers"?
Can you post your domain.xml file?
Does your application contain a JSF implementation (mojarra, icefaces, etc.) ?

**Comment by rds76** [comment]
We are using PrimeFaces. 
```
<http-service>
        <access-log></access-log>
        <virtual-server id="server"></virtual-server>
        <virtual-server id="__asadmin" network-listeners="admin-listener"></virtual-server>
        <virtual-server id="VS1" hosts="www.abc1.com,abc1.com" default-web-module="m1" network-listeners="http-listener-1,http-listener-2"></virtual-server>
        <virtual-server id="VS2" default-web-module="m2" hosts="www.abc2.com" network-listeners="http-listener-1,http-listener-2"></virtual-server>
      </http-service>
```
Can please explain: 
> Have you turned off class loader delegation?

**Comment by mikecroft** [comment]
By default, Payara Server delegates classloading to the application server and therefore libraries from the server are preferred over those from your application. This can be a problem with frameworks like Spring where it's important to use the version that you've built your application around.

If you set the system property `fish.payara.classloading.delegate` to `false`, then you can prefer your own libraries over the ones that come with Payara Server.

Assuming you have packaged PrimeFaces within a WAR archive, you can add
`<class-loader delegate="false"/>` in your `WEB-INF/glassfish-web.xml` which will enable the feature. This was, incidentally, available within GlassFish too - what we have added is the ability to do this with an EAR and across the domain.

https://payara.gitbooks.io/payara-server/content/documentation/extended-documentation/classloading.html#4-solutions-to-the-class-loading-issue

**Comment by mulderbaba** [comment]
@rds76 which version of Payara are you using? Are you deploying more than one JSF + PrimeFaces based application? Are you doing any calls to class javax.faces.FactoryFinder from your application code?

**Comment by rds76** [comment]
@mikecroft I set sevrver-config/System properties `fish.payara.classloading.delegate` to `false`. I doesn't help.
Suspection pointed me Virtual servers as during testing with just single default virt. server `server` we had not seen this warning. It prolly apears when we added 2 another virtual servers.
@mulderbaba Payara 163. We had 2 jsf apps but we tried undeploy and left just one with no success. We have not found any `FactoryFinder` calls.

**Comment by fturizo** [comment]
@rds76, is this issue still persisting in your environment? If so, can you post the _web.xml_ and _faces-config.xml_ of the applications to see if we can reproduce the issue?

**Comment by rds76** [comment]
we turned logging of javax.face to severe to avoid flooding, but assume warning is still there. Which part of config are you interested as it is part of big application unpossible to share

**Comment by fturizo** [comment]
@rds76, I think that with these files we can get a good grasp of a reproducible scenario:

* _faces-config.xml_
* _web.xml_
* _glassfish-web.xml_
* _domain-xml_

If possible, since these are big applications, try to post some sample **JSF** pages and their backing beans if you think they might help reproduce the issue as well.

**Comment by fturizo** [comment]
Hello, @rds76. Is this issue still persisting on your environment? If so can you attach the files I suggested to create a reproducible test case?

**Comment by hberton** [comment]
I have the same problem when I add a new virtual server besides the "server" and "_asadmin"

**Comment by lprimak** [comment]
@fturizo @mikecroft you guys can put this on my queue if you decide it's appropriate

**Comment by OndrejM** [comment]
@lprimak Would you try to reproduce the described behavior then, please?

**Comment by hberton** [comment]
Working 100%

```
<http-service>
	<access-log></access-log>
	<virtual-server network-listeners="http-listener-1,http-listener-2" hosts="${com.sun.aas.hostName},cmcontrol.visum.com.br" sso-enabled="true" id="server"></virtual-server>
	<virtual-server network-listeners="admin-listener" id="__asadmin"></virtual-server>
</http-service>
```

Error 
```
<http-service>
	<access-log></access-log>
	<virtual-server network-listeners="http-listener-1,http-listener-2" sso-enabled="true" id="server"></virtual-server>
	<virtual-server network-listeners="admin-listener" id="__asadmin"></virtual-server>
	<virtual-server network-listeners="http-listener-1,http-listener-2" hosts="test.com" sso-enabled="true" id="test" sso-cookie-http-only="false"></virtual-server>
</http-service>
```
server.log
`Multiple JSF Applications found on same ClassLoader.  Unable to safely determine which FactoryManager instance to use. Defaulting to first match.`

**Comment by z-mathe** [comment]
Hi Fabio, Mike, Ondrej,

This is also the case with us (multiple virtual servers) on Payara v4.1.1.171, Oracle JDK 8u121, windows.
We have got multiple war with primefaces usage, but removed the previously bundled javax.faces.jar from all, so there's only one exists in the \glassfish\modules\ folder (there was a full server restart too).

[#|2017-04-26T18:26:42.356+0200|WARNING|Payara 4.1|javax.faces|_ThreadID=76;_ThreadName=http-thread-pool::jk-connector(4);_TimeMillis=1493224002356;_LevelValue=900;|Multiple JSF Applications found on same ClassLoader.  Unable to safely determine which FactoryManager instance to use. Defaulting to first match.|#]

This warning generated on every webserver call from apache mod_jk sessions, but if no webserver calling and the app become idle, the warning not generated.

**Comment by z-mathe** [comment]
I decompiled the CurrentThreadToServletContext class from \glassfish\modules\javax.faces.jar (ver: 2.2.14, but equals with the new 2.3 version, https://mvnrepository.com/artifact/org.glassfish/javax.faces):

```
private void initFromFactoryMap(ClassLoader cl, Map<FactoryManagerCacheKey, FactoryFinderInstance> factoryMap)
{
  Set<FactoryManagerCacheKey> keys = factoryMap.keySet();
  FactoryManagerCacheKey matchingKey = null;
  
  if (keys.isEmpty()) {
    this.cl = cl;
    this.marker = new Long(System.currentTimeMillis());
  }
  else
  {
    for (FactoryManagerCacheKey currentKey : keys) {
      assert (null != currentKey.cl);
      ClassLoader matchingCL = findMatchConsideringParentClassLoader(cl, currentKey.cl);
      
      if (null != matchingCL)
      {
        if (null != matchingKey) {
          CurrentThreadToServletContext.LOGGER.log(Level.WARNING, "Multiple JSF Applications found on same ClassLoader.  Unable to safely determine which FactoryManager instance to use. Defaulting to first match.");
          break;
        }
        matchingKey = currentKey;
        this.cl = matchingCL;
      }
    }
    if (null != matchingKey) {
      this.marker = matchingKey.marker;
      this.context = matchingKey.context;
    }
  }
}

```

It would be good to know, what keys are in the FactoryManager cache in our environments.
An interesting comment in this "closed" JSF bug report: https://java.net/jira/browse/JAVASERVERFACES-2029, but the attached method patch (multiple-factory-manager.patch) has no trace in the latest implementations.
Guys? Any idea?

**Comment by OndrejM** [comment]
I was able to reproduce with an app that contains JSF servlet mapping in web.xml and is deployed to multiple virtual servers. We'll investigate how we can fix the warning.

**Comment by AllenConquest** [comment]
Is this fix available in the community version of Payara 4 or the customer version?

**Comment by smillidge** [comment]
We don't provide community releases of Payara 4 although all fixes are in the Payara 4 branch on github and this fixed went in with PR #2554

We provide supported binaries for Payara 4 for customers and will continue to do so until 2024 along with support for JDK 8 via support for Zulu Open JDK.
